### PR TITLE
Provide fully initialized route progress from the navigator

### PR DESCRIPTION
**Pull request by LukasPaczos**
Refs https://github.com/mapbox/mapbox-navigation-android/pull/2359 and https://github.com/mapbox/mapbox-navigation-android/pull/2360.

This PR provides a fully initialized route progress object.

Tailwork that we can target here or preferably in a separate PR:
- [ ] Add geojson core dependency to get a polyline precision constant
- [x] Implement `routeGeometryWithBuffer` field
- [x] Check if `CurrentLegAnnotation`s are used and whether we should include them in the progress.
- [ ] Verify if we want to pass `MaxSpeed` in the route progress.

### CODE REVIEW

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt [hunk 1]
```diff
@@ -119,7 +142,166 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
...
+                    // todo mapbox java issue - leg distance is nullable
+                    val distanceTraveled =
+                        (currentLeg.distance()?.toFloat() ?: 0f) - remainingLegDistance
+                    legProgressBuilder.distanceTraveled(distanceTraveled)
+                    legProgressBuilder.fractionTraveled(
+                        distanceTraveled / (currentLeg.distance()?.toFloat() ?: 0f)
+                    )
+
+                    var routeDistanceRemaining = remainingLegDistance
+                    if (legs.size >= 2) {
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt]: Magic number

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt [hunk 2]
```diff
@@ -119,7 +142,166 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
...
+                        for (i in legIndex + 1 until legs.size) {
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt]: Magic number

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt [hunk 3]
```diff
@@ -119,7 +142,166 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
...
+        stepProgressBuilder.distanceRemaining(remainingStepDistance)
+        stepProgressBuilder.durationRemaining((remainingStepDuration / ONE_SECOND_IN_MILLISECONDS).roundToLong())
+
+        legProgressBuilder.currentStepProgress(stepProgressBuilder.build())
+        legProgressBuilder.distanceRemaining(remainingLegDistance)
+        legProgressBuilder.durationRemaining((remainingLegDuration / ONE_SECOND_IN_MILLISECONDS).roundToLong())
+
+        routeProgressBuilder.currentLegProgress(legProgressBuilder.build())
+        if (routeState != RouteState.OFFROUTE) {
+            val state = RouteProgressStateMap[routeState]!!
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt]: Should we remove this `!!`?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteLegProgress.kt [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+class RouteLegProgress private constructor(
+    private val legIndex: Int = 0,
+    private val routeLeg: RouteLeg? = null,
+    private val distanceTraveled: Float = 0f,
+    private val distanceRemaining: Float = 0f,
+    private val durationRemaining: Long = 0L,
+    private val fractionTraveled: Float = 0f,
+    private val currentStepProgress: RouteStepProgress? = null,
+    private val upComingStep: LegStep? = null,
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteLegProgress.kt]: NIT Typo

```suggestion
    private val upcomingStep: LegStep? = null,
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteLegProgress.kt [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+     */
+    fun currentStepProgress(): RouteStepProgress? = currentStepProgress
+
+    /**
+     * Get the next/upcoming step immediately after the current step. If the user is on the last step
+     * on the last leg, this will return null since a next step doesn't exist.
+     *
+     * @return a [LegStep] representing the next step the user will be on.
+     */
+    fun upComingStep(): LegStep? = upComingStep
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteLegProgress.kt]: Typo

```suggestion
    fun upcomingStep(): LegStep? = upComingStep
```

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteLegProgress.kt [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    data class Builder(
+        private var legIndex: Int = 0,
+        private var routeLeg: RouteLeg? = null,
+        private var distanceTraveled: Float = 0f,
+        private var distanceRemaining: Float = 0f,
+        private var durationRemaining: Long = 0L,
+        private var fractionTraveled: Float = 0f,
+        private var currentStepProgress: RouteStepProgress? = null,
+        private var upComingStep: LegStep? = null
```
Comment (Guardiola31337) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/trip/model/RouteLegProgress.kt]: Typo

```suggestion
        private var upcomingStep: LegStep? = null
```

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt [hunk 7]
```diff
@@ -21,7 +39,16 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
...
+    private const val ONE_INDEX = 1
+    private const val ONE_SECOND_IN_MILLISECONDS = 1000.0
+    private const val FIRST_BANNER_INSTRUCTION = 0
+    private const val GRID_SIZE = 0.0025f
+    private const val BUFFER_DILATION: Short = 1
+    private const val TWO_LEGS: Short = 2
+
     private val navigator: Navigator = Navigator()
+    private var route: DirectionsRoute? = null
+    private var routeBufferGeojson: Geometry? = null
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt]: Typo

```suggestion
    private var routeBufferGeoJson: Geometry? = null
```

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt [hunk 8]
```diff
@@ -120,6 +157,172 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
...
     private fun NavigationStatus.getRouteProgress(): RouteProgress {
-        return RouteProgress("")
+        val upcomingStepIndex = stepIndex + ONE_INDEX
+
+        val routeProgressBuilder = RouteProgress.Builder()
+        val legProgressBuilder = RouteLegProgress.Builder()
+        val stepProgressBuilder = RouteStepProgress.Builder()
+
+        ifNonNull(route) { route ->
+            ifNonNull(route.legs()) { legs ->
```
Comment (Lebedinsky) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/MapboxNativeNavigatorImpl.kt]: we can join this two checks with one
```suggestion
            ifNonNull(route?.legs()) { legs ->
```

**Pull request merged**